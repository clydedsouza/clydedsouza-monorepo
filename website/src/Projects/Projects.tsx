export interface IProjectProps {
  name: string
}

function Projects(props: IProjectProps) {
  return (
    <div>
      <p>Project {props.name}</p>
      <p>ifbwebfwebfi</p>
    </div>
  )
}

export default Projects
