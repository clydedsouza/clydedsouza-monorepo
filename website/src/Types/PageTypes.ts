export enum PageTypes {
  About = 'About',
  Projects = 'Projects',
  Speaking = 'Speaking',
  Teaching = 'Teaching',
  Books = 'Books',
  Highlights = 'Highlights',
  Pinned = 'Pinned',
}
