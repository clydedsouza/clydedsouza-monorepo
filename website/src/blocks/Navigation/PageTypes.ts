export enum PageTypes {
  About = "About",
  Portfolio = "Portfolio",
  Platforms = "Platforms",
  Highlights = "Highlights",
  Cta = "Cta",
  Blog = "Blog",
}
