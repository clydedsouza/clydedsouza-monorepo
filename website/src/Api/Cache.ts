import { PageTypes } from '../Types/PageTypes'
import { IProjectData } from '../Types/ProjectData'
import { getProjectData } from './Core'

const projectDataCacheMap: Map<PageTypes, Promise<IProjectData>> = new Map()

export const getCachedProjectData = (
  pageType: PageTypes
): Promise<IProjectData> => {
  if (!projectDataCacheMap.has(pageType)) {
    const getProjectDataPromise = getProjectData(pageType).catch((error) => {
      projectDataCacheMap.delete(pageType)
      return Promise.reject(error)
    })
    projectDataCacheMap.set(pageType, getProjectDataPromise)
  }

  return projectDataCacheMap.get(pageType)!
}
