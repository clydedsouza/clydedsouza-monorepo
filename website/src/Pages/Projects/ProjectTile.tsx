import { IProject } from '../../Types/ProjectData'

function ProjectTile(props: IProject) {
  return (
    <section>
      <div
        className="project-header"
        style={{ backgroundImage: `url(${props.image})` }}
      >
        <div className="project-header-contents">
          <div>
            <span>{props.category}</span>
            {!props.isActive && <span>Inactive</span>}
          </div>
        </div>
      </div>
      <div className="project-body">
        <h3>{props.title}</h3>
        <p>{props.description}</p>
        <p>
          {props.madeUsing &&
            props.madeUsing.map((item) => <span key={item}>{item}</span>)}
        </p>
        <div>
          <div className="col-xs-8 text-left">&nbsp;</div>
          <div className="col-xs-4 text-right">
            {props.isActive && props.hasWebsite && (
              <a href={props.website} target="_blank" rel="noreferrer">
                <i className="fas fa-globe"></i>
              </a>
            )}
            {props.onGithub && (
              <a href={props.github} target="_blank" rel="noreferrer">
                <i className="fab fa-github"></i>
              </a>
            )}
          </div>
        </div>
      </div>
    </section>
  )
}

export default ProjectTile
