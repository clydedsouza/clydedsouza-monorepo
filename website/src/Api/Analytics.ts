import mixpanel from 'mixpanel-browser'
import { PageTypes } from '../Components/Navigation/PageTypes'
import {
  ILinkClickedProps,
  MIXPANEL_DEV_PROJECT_ID,
  MIXPANEL_PROD_PROJECT_ID,
} from './IAnalytics'

export const initAnalyticsWithSuperProperties = () => {
  const mixpanelProjectId =
    process.env.NODE_ENV === 'production'
      ? MIXPANEL_PROD_PROJECT_ID
      : MIXPANEL_DEV_PROJECT_ID

  try {
    mixpanel.init(mixpanelProjectId)
    mixpanel.register({
      environment: process.env.NODE_ENV,
    })
  } catch {
    undefined
  }
}

export const sendPageViewEvent = (pageType: PageTypes) => {
  try {
    mixpanel.track('Page view', { page: pageType })
  } catch {
    undefined
  }
}

export const sendLinkClickedEvent = (props: ILinkClickedProps) => {
  try {
    mixpanel.track('Link clicked', {
      destination_url: props.link,
      link_type: props.type,
      page_location: props.location,
    })
  } catch {
    undefined
  }
}
