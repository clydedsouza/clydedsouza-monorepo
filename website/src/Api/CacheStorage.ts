import { PageTypes } from '../Components/Navigation/PageTypes'
import { IProjectData } from './IProjectData'

export const projectDataCacheMap: Map<PageTypes, IProjectData> = new Map()
