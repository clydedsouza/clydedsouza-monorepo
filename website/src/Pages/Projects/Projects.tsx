import { useEffect, useState } from 'react'
import { getCachedProjectData } from '../../Api/Cache'
import Seo from '../../Components/Seo/Seo'
import { PageTypes } from '../../Types/PageTypes'
import { IProject } from '../../Types/ProjectData'
import ProjectTile from './ProjectTile'

export interface IProjectProps {
  name: PageTypes
}

function Projects(props: IProjectProps) {
  const [projectData, setProjectData] = useState<IProject[]>([])
  const [isLoading, setIsLoading] = useState<boolean>(true)
  const [showInactive, toggleShowInactive] = useState<boolean>(false)

  useEffect(() => {
    setIsLoading(true)
    getCachedProjectData(props.name)
      .then((apiResponse) => {
        setProjectData(apiResponse.data)
        setIsLoading(false)
      })
      .catch(() => {
        setProjectData([])
      })
  }, [props.name])

  return (
    <>
      <Seo title={props.name} />
      {isLoading ? (
        <p>Contents are loading...</p>
      ) : (
        <>
          <form className="form-inline">
            <div className="noinactive hidden" />
            <input
              type="search"
              id="searchTxt"
              className="form-control"
              placeholder="Search..."
            />
            <div id="includeInactiveContainer">
              <label className="switch-label" htmlFor="includeInactive">
                Show inactive
              </label>
              <label className="switch">
                <input
                  type="checkbox"
                  id="includeInactive"
                  onClick={() => toggleShowInactive(!showInactive)}
                />
                <span className="slider round"></span>
              </label>
            </div>
          </form>
          <div id="pinnedProjects">
            <div id="pinnedProjectsList">
              {projectData.length > 0 ? (
                projectData
                  .sort((dateA, dateB) => {
                    return (
                      Number(new Date(dateB.date)) -
                      Number(new Date(dateA.date))
                    )
                  })
                  .map((item) => <ProjectTile {...item} key={item.id} />)
              ) : (
                <p>Couldn&lsquo;t load the data at this stage</p>
              )}
            </div>
          </div>
        </>
      )}
    </>
  )
}

export default Projects
