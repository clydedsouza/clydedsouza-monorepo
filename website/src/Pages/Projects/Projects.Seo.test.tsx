import {
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved,
} from '@testing-library/react'
import * as apiCacheModule from '../../Api/Cache'
import { PageTypes } from '../../Types/PageTypes'
import Projects from './Projects'

jest.mock('../../Api/Cache')

describe('Projects', () => {
  beforeEach(() => {
    jest
      .spyOn(apiCacheModule, 'getCachedProjectData')
      .mockImplementationOnce(() =>
        Promise.resolve({
          app: {},
          data: [],
        })
      )
  })

  describe.each([PageTypes.Highlights, PageTypes.Platforms])(
    'given page type is %s',
    (pageType: PageTypes) => {
      it(`should render projects seo`, async () => {
        render(<Projects {...{ name: pageType }} />)
        await waitForElementToBeRemoved(() =>
          screen.queryByText('Contents are loading...')
        )
        await waitFor(() =>
          expect(document.title).toEqual(
            `${pageType} – Clyde D'Souza – Software Engineer and Author`
          )
        )
        expect(document.head).toMatchSnapshot()
      })
    }
  )
})
