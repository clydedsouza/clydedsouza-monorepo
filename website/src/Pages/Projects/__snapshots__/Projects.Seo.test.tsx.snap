// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects given page type is Highlights should render projects seo 1`] = `
<head>
  <title>
    Highlights – Clyde D'Souza – Software Engineer and Author
  </title>
  <link
    data-react-helmet="true"
    href="https://clydedsouza.net/highlights"
    rel="canonical"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-react-helmet="true"
    name="description"
  />
  <meta
    content="https://clydedsouza.net/highlights"
    data-react-helmet="true"
    property="og:url"
  />
  <meta
    content="Highlights – Clyde D'Souza – Software Engineer and Author"
    data-react-helmet="true"
    property="og:title"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-react-helmet="true"
    property="og:description"
  />
  <meta
    content="https://clydedsouza.net/images/clydedsouza-share-2023.jpg"
    data-react-helmet="true"
    property="og:image"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-react-helmet="true"
    property="twitter:card"
  />
  <meta
    content="Highlights – Clyde D'Souza – Software Engineer and Author"
    data-react-helmet="true"
    property="twitter:title"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-react-helmet="true"
    property="twitter:description"
  />
  <meta
    content="https://clydedsouza.net/images/clydedsouza-share-2023.jpg"
    data-react-helmet="true"
    property="twitter:image"
  />
</head>
`;

exports[`Projects given page type is Platforms should render projects seo 1`] = `
<head>
  <title>
    Platforms – Clyde D'Souza – Software Engineer and Author
  </title>
  <link
    data-react-helmet="true"
    href="https://clydedsouza.net/platforms"
    rel="canonical"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-react-helmet="true"
    name="description"
  />
  <meta
    content="https://clydedsouza.net/platforms"
    data-react-helmet="true"
    property="og:url"
  />
  <meta
    content="Platforms – Clyde D'Souza – Software Engineer and Author"
    data-react-helmet="true"
    property="og:title"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-react-helmet="true"
    property="og:description"
  />
  <meta
    content="https://clydedsouza.net/images/clydedsouza-share-2023.jpg"
    data-react-helmet="true"
    property="og:image"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-react-helmet="true"
    property="twitter:card"
  />
  <meta
    content="Platforms – Clyde D'Souza – Software Engineer and Author"
    data-react-helmet="true"
    property="twitter:title"
  />
  <meta
    content="Clyde D'Souza is software engineer and author based in Auckland, New Zealand. Feel free to send him a tweet @ClydeDz"
    data-react-helmet="true"
    property="twitter:description"
  />
  <meta
    content="https://clydedsouza.net/images/clydedsouza-share-2023.jpg"
    data-react-helmet="true"
    property="twitter:image"
  />
</head>
`;
